src/persist-client/src/collection.rs

```diff
@@ -0,0 +1,112 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::sync::Arc;
+use std::time::Instant;
+
+use anyhow::anyhow;
+use mz_persist::s3::S3BlobMultiWriter;
+use mz_persist::storage::{Atomicity, StorageError};
+use uuid::Uuid;
+
+use crate::metadata::CollectionMeta;
+use crate::paths::Paths;
+use crate::{CompareAndSet, Id, Log, MemLog, SeqNo};
+
+#[derive(Debug, Clone)]
+pub struct Collection {
+    pub(crate) id: Id,
+    pub(crate) blob: S3BlobMultiWriter,
+    pub(crate) log: Arc<MemLog>,
+}
+
+impl Collection {
+    pub async fn update_metadata<
+        R,
+        F: FnMut(SeqNo, &Option<CollectionMeta>) -> (CollectionMeta, R),
+    >(
+        &mut self,
+        deadline: Instant,
+        mut f: F,
+    ) -> Result<(Option<CollectionMeta>, CollectionMeta, SeqNo, R), StorageError> {
+        loop {
+            let (prev_seqno, prev_meta) = self.fetch_meta(deadline).await?;
+            let (new_meta, ret) = f(prev_seqno, &prev_meta);
```

Comment:
if this return value is an error do we still want to try to write out the meta and update log?

Reply:
hmm this is a good question. at the moment, `update_metadata` only knows that it's getting an R and blindly passes it back at the end. i agree that it's surprising to a code reader to see a call of this method get a closure that returns a Result and not have it do what you say. i can't say offhand whether we want to force the closure to return a Result or not. more thinking to do here

Reply:
I think at a higher level - it was confusing to me to see `f` as a function from old_meta to new_meta that was infallible because that seemed too strong to be true - that we could _always_ take a meta, apply some `f` to it and get a new meta back. Then I looked at the usage and saw R was a Result, which made more sense

Reply:
yeah i think that's compelling